tests/ui/redundant_type_annotations.rs

```diff
@@ -0,0 +1,151 @@
+#![allow(unused)]
+#![warn(clippy::redundant_type_annotations)]
+
+#[derive(Debug, Default)]
+struct Cake<T> {
+    _data: T,
+}
+
+fn make_something<T: Default>() -> T {
+    T::default()
+}
+
+fn make_cake<T: Default>() -> Cake<T> {
+    Cake::<T>::default()
+}
+
+fn plus_one<T: std::ops::Add<u8, Output = T>>(val: T) -> T {
+    val + 1
+}
+
+struct Pie {
+    inner: u32,
+}
+
+enum Pizza {
+    One,
+    Two,
+}
+
+fn return_a_string() -> String {
+    String::new()
+}
+
+fn return_a_struct() -> Pie {
+    Pie { inner: 5 }
+}
+
+fn return_an_enum() -> Pizza {
+    Pizza::One
+}
+
+fn return_an_int() -> u32 {
+    5
+}
+
+impl Pie {
+    fn return_an_int(&self) -> u32 {
+        self.inner
+    }
+
+    fn associated_return_an_int() -> u32 {
+        5
+    }
+
+    fn new() -> Self {
+        Self { inner: 5 }
+    }
+
+    fn associated_return_a_string() -> String {
+        String::from("")
+    }
+
+    fn test_method_call(&self) {
+        let v: u32 = self.return_an_int(); // Should lint
+    }
+}
+
+fn test_generics() {
+    // The type annotation is needed to determine T
+    let _c: Cake<i32> = make_something();
+
+    // The type annotation is needed to determine the topic
+    let _c: Cake<u8> = make_cake();
+
+    // This should lint (doesn't)
+    let _c: Cake<u8> = make_cake::<u8>();
+
+    // This should lint (doesn't)
+    let _c: u8 = make_something::<u8>();
```

Comment:
These are false negatives, which are fine to have in my opinion. I'd suggest merging the lint as is, we can always improve it later. Though, the comment should make it a bit clearer then. Same for the cases below.

Maybe:

```suggestion
    // This could be lint, but currently isn't 
    let _c: Cake<u8> = make_cake::<u8>();

    // This could be lint, but currently isn't
    let _c: u8 = make_something::<u8>();
```